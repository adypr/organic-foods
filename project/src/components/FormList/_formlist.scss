.form-list {
    width: 40%;
   min-width: 450px;
   margin-top: 5px;
   padding-left: 10px;;
   border-left: 1px solid rgba($color: #000000, $alpha: 0.5);

   &__title {
    margin-top: 14px;
   }
}