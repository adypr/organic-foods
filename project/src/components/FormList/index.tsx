import React from 'react';
import Product from 'components/Product';
import { CardsContent } from 'base/types';

import './_formlist.scss';

const FormList: React.FC<{
  products: CardsContent;
  children?: React.ReactNode;
  onClick: (productId: string) => void;
}> = ({ products, children, onClick }) => {
  const handleProductClick = (e: React.MouseEvent) => {
    if ((e.target as HTMLElement).classList.contains('remove')) onClick(e.currentTarget.id);
  };

  return (
    <div className="form-list">
      <h2 className="form-list__title">Your goods</h2>
      <div className="form-list__items">
        {!products.length ? (
          <span>There&apos;re no items</span>
        ) : (
          products.map((product, i) => {
            return <Product key={i} {...product} onClick={handleProductClick} />;
          })
        )}
        {children}
      </div>
    </div>
  );
};

export default FormList;
