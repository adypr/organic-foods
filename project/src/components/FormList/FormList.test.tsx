import React from 'react';
import { render, screen } from '@testing-library/react';

import FormList from './index';
import userEvent from '@testing-library/user-event';

const mockFields = [
  {
    title: 'Title',
    currentPrice: '100',
    oldPrice: '200',
    img: 'file.png',
    description: 'Product description',
    bestBefore: '2024-04-04',
    productType: 'resale',
    promo: 'promoYes',
    delivery: 'post',
    id: '123',
  },
];

describe('FormList', () => {
  const mockFn = jest.fn();

  test('matches snapshot', () => {
    const formList = render(<FormList products={mockFields} onClick={mockFn} />);
    expect(formList).toMatchSnapshot();
  });

  test("doesn't renders any product item after initialization", () => {
    render(<FormList products={[]} onClick={mockFn} />);
    expect(screen.queryByTestId('productTitle')).not.toBeInTheDocument();
  });

  test('adds new product', () => {
    render(<FormList products={mockFields} onClick={mockFn} />);
    expect(screen.getByTestId('productTitle')).toBeInTheDocument();
  });

  test('call a function when the delete button is clicked', () => {
    render(<FormList products={mockFields} onClick={mockFn} />);
    userEvent.click(screen.getByTestId('closeButton'));
    expect(mockFn).toBeCalled();
  });

  test('renders message if list is empty', () => {
    render(<FormList products={[]} onClick={mockFn} />);
    expect(screen.getByText(/there're no items/i)).toBeInTheDocument();
  });

  test('renders message after deleting last product', () => {
    render(<FormList products={mockFields} onClick={mockFn} />);
    userEvent.click(screen.getByTestId('closeButton'));
    expect(screen.getByText(/there're no items/i)).toBeInTheDocument();
  });
});
