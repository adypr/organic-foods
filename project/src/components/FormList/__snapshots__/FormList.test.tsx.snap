// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormList matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-list"
      >
        <h2
          class="form-list__title"
        >
          Your goods
        </h2>
        <div
          class="form-list__items"
        >
          <div
            class="product"
            data-testid="productTitle"
            id="123"
          >
            <img
              alt="Title"
              class="product__img"
              height="100px"
              src="file.png"
              width="100px"
            />
            <div
              class="product__content"
            >
              <h4
                class="product__title"
              >
                Title
              </h4>
              <div
                class="product__prices"
              >
                <span
                  class="product__price"
                >
                  Current price: 
                  100
                </span>
                <span
                  class="product__old-price"
                >
                  Old price: 
                  200
                </span>
              </div>
              <div
                class="product__settings"
              >
                <div
                  class="product__setting product__type"
                >
                  type: 
                  resale
                </div>
                <div
                  class="product__setting product__action"
                >
                  promo: 
                  âœ…
                </div>
                <div
                  class="product__setting product__date"
                >
                  best before: 
                  2024-04-04
                </div>
                <div
                  class="product__setting product__delivery"
                >
                  delivery: 
                  post
                </div>
              </div>
              <details
                class="product__description"
              >
                <summary>
                  Description
                </summary>
                <div
                  class="product__text"
                  data-testid="productDescription"
                >
                  Product description
                </div>
              </details>
            </div>
            <ul
              class="product__control"
            >
              <li
                class="control__item"
              >
                <a
                  class="control__link edit"
                  href="#"
                >
                  ðŸ–‰
                </a>
              </li>
              <li
                class="control__item"
              >
                <a
                  class="control__link remove"
                  data-testid="closeButton"
                  href="#"
                >
                  âœ–
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-list"
    >
      <h2
        class="form-list__title"
      >
        Your goods
      </h2>
      <div
        class="form-list__items"
      >
        <div
          class="product"
          data-testid="productTitle"
          id="123"
        >
          <img
            alt="Title"
            class="product__img"
            height="100px"
            src="file.png"
            width="100px"
          />
          <div
            class="product__content"
          >
            <h4
              class="product__title"
            >
              Title
            </h4>
            <div
              class="product__prices"
            >
              <span
                class="product__price"
              >
                Current price: 
                100
              </span>
              <span
                class="product__old-price"
              >
                Old price: 
                200
              </span>
            </div>
            <div
              class="product__settings"
            >
              <div
                class="product__setting product__type"
              >
                type: 
                resale
              </div>
              <div
                class="product__setting product__action"
              >
                promo: 
                âœ…
              </div>
              <div
                class="product__setting product__date"
              >
                best before: 
                2024-04-04
              </div>
              <div
                class="product__setting product__delivery"
              >
                delivery: 
                post
              </div>
            </div>
            <details
              class="product__description"
            >
              <summary>
                Description
              </summary>
              <div
                class="product__text"
                data-testid="productDescription"
              >
                Product description
              </div>
            </details>
          </div>
          <ul
            class="product__control"
          >
            <li
              class="control__item"
            >
              <a
                class="control__link edit"
                href="#"
              >
                ðŸ–‰
              </a>
            </li>
            <li
              class="control__item"
            >
              <a
                class="control__link remove"
                data-testid="closeButton"
                href="#"
              >
                âœ–
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
